### PR TITLE
[DNM/WIP]Projectile Dampening Fields for Peacekeeper Cyborgs

### PR DESCRIPTION
**Pull request by silicons**
TL;DR: Projectile dampening field for PKborgs.
Reduces damage by 50%
Has 2000 max energy
Takes 1 energy unit from cyborg cell to recharge 2 field energy
Recharges 100 field energy/second
each point of damage reduced drains 20 energy/second (Lasers would drain 200 energy/second)
Projectiles are slowed by 50% 
Each projectile in the field being slowed will drain 100 energy/second
Field is 7x7 centered on the cyborg
Field will slow cyborg down by 1 while active (Not coded yet)
Projectiles are restored upon exiting the field.
However: Projectiles fired INSIDE the field will be unaffected, until they exit the field, at which they're permanently weakened so you can't use the field as running cover while firing out of it.
There's visual effects for the edges of the field as well as projectiles being affected by it

**Adds /datum/field, meant to be used for movable AOE effects that are hooked into turf's Enter(). I'll be using them after this PR for things like shield generators and whatnot, so yes, there are uses for it and I didn't just use hardcoded fields in the dampening fields for a reason.**

:cl:
rscadd: Peacekeeper cyborgs now have projectile dampening fields. Beware of its high energy consumption...
/:cl:

WORKING DEMO

### CODE REVIEW

File: tgstation.dme [hunk 1]
```diff
@@ -672,10 +673,10 @@
 #include "code\game\objects\effects\decals\decal.dm"
 #include "code\game\objects\effects\decals\misc.dm"
 #include "code\game\objects\effects\decals\remains.dm"
-#include "code\game\objects\effects\decals\cleanable\aliens.dm"
-#include "code\game\objects\effects\decals\cleanable\humans.dm"
-#include "code\game\objects\effects\decals\cleanable\misc.dm"
-#include "code\game\objects\effects\decals\cleanable\robots.dm"
+#include "code\game\objects\effects\decals\Cleanable\aliens.dm"
```
Comment (kevinz000) [inline tgstation.dme]: note to self to fix my shit

**Comment by kevinz000** [comment]
also i'd make them reduce melee damage too but uh atomized PRs and it'll probably get different feedback.

**Comment by Repukan** [comment]
>Reducing Ion effectiveness 

Shit, the ion bolts aren't even that fast either.

**Comment by kevinz000** [comment]
@Repukan sure i'll whitelist ion out since they're special and would probably break shielding anyways

**Comment by kevinz000** [comment]
wait they're already not affected because they're nodamage.

File: code/modules/fields/fields.dm [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+	var/list/turf/edge_turfs_new = list()
+
+/datum/field/New()
+	SSfields.register_new_field(src)
+	..()
+
+/datum/field/Destroy()
+	SSfields.unregister_field(src)
+	full_cleanup()
+	..()
```
Comment (ChangelingRain) [inline code/modules/fields/fields.dm]: return ..()

File: code/modules/fields/fields.dm [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+	var/setup_field_turfs = FALSE
+
+	var/list/turf/field_turfs = list()
+	var/list/turf/edge_turfs = list()
+	var/list/turf/field_turfs_new = list()
+	var/list/turf/edge_turfs_new = list()
+
+/datum/field/New()
+	SSfields.register_new_field(src)
+	..()
```
Comment (ChangelingRain) [inline code/modules/fields/fields.dm]: don't call ..() here

**Comment by InsaneHyena** [comment]
Peacekeepers must remain useless.

**Comment by Militaires** [comment]
pics needed

**Comment by GunHog** [comment]
Makes Peacekeepers a hilarious inhibitor to combat! I would like to see it in action.

**Comment by Repukan** [comment]
Actually yeah aren't peacekeepers supposed to be useless?

**Comment by GunHog** [comment]
They use confusion and other annoying abilities decrease effectiveness in combat without halting it. I have thought of them as changing battlefield dynamics. At the very lease, they are good for grabbing stunned or crit humans.

**Comment by Repukan** [comment]
Any borg is good for grabbing humans, medbots better since they can actually heal.

Yeah I'm not sure if giving them this actually accomplishes their goal of being a middle finger to secborg sympathizers

**Comment by ExcessiveUseOfCobblestone** [comment]
They aren't suppose to be useless, they're suppose to be good at keeping the peace without direct combat modules like a GG stun devices [flash is the exception :^)]

**Comment by Robustin** [comment]
Lower the recharge rate it seems far too high. You'd pretty much have to constantly be flinging projectiles at the borg to actually keep the field down for any meaningful period of time.

**Comment by PKPenguin321** [comment]
Pics?

**Comment by Bawhoppen** [comment]
peacekeepers not a disabler and zipties tbh

**Comment by GunHog** [comment]
> peacekeepers need a disabler and zipties tbh

That is called "Security Cyborg"

**Comment by kevinz000** [comment]
Pics are inbound once I get the graphics fully working
@Robustin I'm still tweaking values, this is WIP :P

**Comment by Robustin** [comment]
Which is why im suggesting that you tweak values instead of giving you a thumbs down and posting

"SHIT IDEA BORGS GONNA HAVE INFINITE PROJECTILE SLOWING FUCKING BROKEN"

**Comment by Bawhoppen** [comment]
@GunHog reee Poe's Law

**Comment by InsaneHyena** [comment]
Peacekeepers were originally intended to mock silicon babies, who cried about removal of secborgs. Giving them useful stuff defeats the point.

**Comment by kevinz000** [comment]
right gonna have to wait for pictures, i'm having a shit time making overlays work :/

**Comment by kevinz000** [comment]
https://gyazo.com/4c50b9b5dfcea8cfdc8f5cc8c158aefd

**Comment by BeeSting12** [comment]
How are stunning projectiles affected by this

**Comment by kevinz000** [comment]
@BeeSting12 they're not . they're not harmful why would the borg be interested in stopping it?1

**Comment by kevinz000** [comment]
actually that might be bad, since it'd be a buff to stun combat. should stuns be slowed/weakened, y/n?

**Comment by GunHog** [comment]
Y

File: code/game/objects/items/robot/robot_items.dm [hunk 4]
```diff
@@ -504,6 +504,144 @@
...
+
+*/
+
+/obj/item/borg/projectile_dampen/debug
+	maxenergy = 50000
+	energy = 50000
+	energy_recharge = 5000
+
+/obj/item/borg/projectile_dampen/Initialize()
+	..()
```
Comment (ChangelingRain) [inline code/game/objects/items/robot/robot_items.dm]: . = ..()

File: code/game/objects/items/robot/robot_items.dm [hunk 5]
```diff
@@ -504,6 +504,144 @@
...
+	projectile_effect = image('icons/effects/fields.dmi', "projectile_dampen_effect")
+	tracked = list()
+	icon_state = "shield0"
+
+/obj/item/borg/projectile_dampen/New()
+	..()
+	START_PROCESSING(SSfastprocess, src)
```
Comment (ChangelingRain) [inline code/game/objects/items/robot/robot_items.dm]: why do you have both an Initialize() and a New() put this in Initialize()

File: code/game/objects/items/robot/robot_items.dm [hunk 6]
```diff
@@ -504,6 +504,144 @@
...
+
+/obj/item/borg/projectile_dampen/Destroy()
+	STOP_PROCESSING(SSfastprocess, src)
+	..()
```
Comment (ChangelingRain) [inline code/game/objects/items/robot/robot_items.dm]: return ..()

File: code/game/objects/items/robot/robot_items.dm [hunk 7]
```diff
@@ -504,6 +504,144 @@
...
+		host.cell.use(energy_recharge*energy_recharge_cyborg_drain_coefficient)
+		energy += energy_recharge
+
+/obj/item/borg/projectile_dampen/proc/dampen_projectile(obj/item/projectile/P, track_projectile = TRUE)
+	if(P in tracked)
+		return
+	if(!P.damage || P.nodamage)
+		return
+	if(track_projectile)
+		tracked += P
```
Comment (ChangelingRain) [inline code/game/objects/items/robot/robot_items.dm]: tracked[P] = TRUE

File: code/game/objects/items/robot/robot_items.dm [hunk 8]
```diff
@@ -504,6 +504,144 @@
...
+/obj/item/borg/projectile_dampen/proc/process_recharge()
+	if(!istype(host))
+		energy = Clamp(energy + energy_recharge, 0, maxenergy)
+		return
+	if(host.cell.charge >= (host.cell.maxcharge * cyborg_cell_critical_percentage))
+		host.cell.use(energy_recharge*energy_recharge_cyborg_drain_coefficient)
+		energy += energy_recharge
+
+/obj/item/borg/projectile_dampen/proc/dampen_projectile(obj/item/projectile/P, track_projectile = TRUE)
+	if(P in tracked)
```
Comment (ChangelingRain) [inline code/game/objects/items/robot/robot_items.dm]: if(tracked[P])

**Review by ChangelingRain — COMMENTED** [review]
this is cheaper

File: code/game/objects/items/robot/robot_items.dm (hunk 8 above)
Comment (RemieRichards) [inline code/game/objects/items/robot/robot_items.dm]: <i>She's learning</i>
